File: Sliding-Window-Maximum.java
1class Solution {
2    class Pair
3{
4    Integer number;
5    Integer index;
6
7    Pair(int number, int index)
8    {
9        this.number = number;
10        this.index = index;
11    }
12    public String toString()
13    {
14        return "Number: "+ number + " Index: "+ index;
15    }
16};
17     public int[] maxSlidingWindow(int[] nums, int k)
18    {
19        Deque<Pair> deque = new ArrayDeque<Pair>();
20        int[] ans = new int[nums.length - k + 1];
21        for(int r=0;r<k;r++)
22        {
23            while( !deque.isEmpty() && nums[r] > deque.peekLast().number)
24            {
25                deque.pollLast();
26            }
27            deque.offerLast(new Pair(nums[r], r));
28        }
29        //System.out.println(deque);
30        ans[0]= deque.peekFirst().number;
31        int startIndex = 0;
32        for(int r=k;r< nums.length;r++)
33        {
34
35            while( !deque.isEmpty() && nums[r] > deque.peekLast().number)
36            {
37                deque.pollLast();
38            }
39
40            deque.offerLast(new Pair(nums[r], r));
41
42            while(deque.peekFirst().index <= startIndex)
43            {
44                deque.pollFirst();
45            }
46            startIndex++;
47            ans[r-k+1] = deque.peekFirst().number;
48
49           // System.out.println(deque);
50        }
51
52
53        return ans;
54
55    }
56}
57

File: Sliding-Window-Maximum.cpp
1class Solution {
2public:
3    vector<int> maxSlidingWindow(vector<int>& nums, int k) 
4    {
5       
6        deque<pair<int,int>> maxCandidates;
7        int max= INT_MIN;
8        vector<int> ans;
9        if(k >= nums.size())
10        {
11            for(int i=0;i<nums.size();i++)
12            {
13                if(max < nums[i])
14                {
15                    max = nums[i];
16                }
17            }
18            ans.push_back(max);
19            return ans;
20        }
21
22        for(int i=0;i<k;i++)
23            {
24                while(!maxCandidates.empty() && maxCandidates.back().first <= nums[i])
25                {
26                    maxCandidates.pop_back();
27                }
28                maxCandidates.push_back({nums[i], i});
29            }
30        ans.push_back(maxCandidates.front().first);
31
32
33        for(int i=1;i<=nums.size()-k;i++)
34        {
35            
36            if(!maxCandidates.empty() && maxCandidates.front().second < i)
37                maxCandidates.pop_front();
38            while(!maxCandidates.empty() && maxCandidates.back().first <= nums[i+ k-1])
39            {
40               // cout<<"Popped " <<maxCandidates.back().first<<endl;
41                maxCandidates.pop_back();
42            }
43            maxCandidates.push_back({nums[i+ k-1],i+ k-1});
44            ans.push_back(maxCandidates.front().first);
45        }
46        return ans;
47        
48    }
49};
50